### PR TITLE
Finished the working concept for the new Danmaku Battles

### LINKED ISSUE #29: Rework the Danmaku Fight System
The Danmaku fight system is a nice bit of flair and fluff, but it has little connection to the rest of the game. Danmaku fights should be more integrated into the mod's gameloop.

### LINKED ISSUE #30: Completely Rework the Incident System
The incidents are incredibly prone to breaking apart and need a big refactoring so it does not cause as many issues anymore as it does right now.

### PR DESCRIPTION
**Pull request by ComradeFlan**
Closes #29 and #30

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,10 @@
+
+Script Code Support/.vs
+Script Code Support/x64
+Script Code Support/Script Code Support/x64
+Script Code Support/Script Code Support/x64
+Script Code Support/Script Code Support/Script Code Support.vcxproj
+Script Code Support/Script Code Support/Script Code Support.vcxproj.filters
+Script Code Support/Script Code Support/Script Code Support.vcxproj.user
+Script Code Support/Script Code Support.sln
```
Comment (Hecatia-Lapislazuli) [inline .gitignore]: If we're going to be tracking some of the code, we should also be tracking the solution file and .vs

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,10 @@
+
+Script Code Support/.vs
+Script Code Support/x64
+Script Code Support/Script Code Support/x64
+Script Code Support/Script Code Support/x64
```
Comment (Hecatia-Lapislazuli) [inline .gitignore]: Double entry

File: common/custom_gui/custom_gui.txt
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/custom_gui.txt:1]: Copied from `example.txt`, should probably either match name or be removed

File: common/custom_gui/th_danmaku_gui.txt [hunk 3]
```diff
@@ -0,0 +1,2121 @@
...
+	frame = { trigger = { th_has_specific_touhou_character = { who = chiyari } } 		number = 148 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = hisami } } 		number = 149 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = zanmu } } 			number = 150 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = rinnosuke } } 		number = 151 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = tokiko } } 		number = 152 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = miyoi } } 			number = 153 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = mizuchi } } 		number = 154 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = zun } } 			number = 155 }
+	frame = { trigger = { th_has_specific_touhou_character = { who = generic } } 		number = 156 }
+	frame = { trigger = { always = yes } 												number = 1 }	#Empty
```
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/th_danmaku_gui.txt:183]: Indentation should probably be adjusted

File: common/custom_gui/th_danmaku_gui.txt [hunk 4]
```diff
@@ -0,0 +1,2121 @@
...
```
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/th_danmaku_gui.txt:646]: Again with indentation

File: common/custom_gui/th_danmaku_gui.txt [hunk 5]
```diff
@@ -0,0 +1,2121 @@
...
+	name = th_danmaku_battle_window_own_class_icon
+	potential = { always = yes }
+	tooltip = th_danmaku_battle_window_own_class_icon_desc
+    frame = {
+		number = 1
+		trigger = { tag = MIM }
+    }
+    frame = {
+        number = 2
+		trigger = { NOT = { tag = MIM } }
```
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/th_danmaku_gui.txt:658]: Indentation

File: common/custom_gui/th_danmaku_gui.txt [hunk 6]
```diff
@@ -0,0 +1,2121 @@
...
+	frame = { trigger = { th_enemy_is_touhou_character = { who = chiyari } } 		number = 148 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = hisami } } 		number = 149 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = zanmu } } 			number = 150 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = rinnosuke } } 		number = 151 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = tokiko } } 		number = 152 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = miyoi } } 			number = 153 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = mizuchi } } 		number = 154 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = zun } } 			number = 155 }
+	frame = { trigger = { th_enemy_is_touhou_character = { who = generic } } 		number = 156 }
+	frame = { trigger = { always = yes } 											number = 1 }	#Empty
```
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/th_danmaku_gui.txt:1376]: Indentation

File: common/custom_gui/th_danmaku_gui.txt [hunk 7]
```diff
@@ -0,0 +1,2121 @@
...
+	name = th_danmaku_battle_window_enemy_class_icon
+	potential = { always = yes }
+	tooltip = th_danmaku_battle_window_enemy_class_icon_desc
+    frame = {
+		number = 1
+		trigger = { tag = MIM }
+    }
+    frame = {
+        number = 2
+		trigger = { NOT = { tag = MIM } }
```
Comment (Hecatia-Lapislazuli) [inline common/custom_gui/th_danmaku_gui.txt:1388]: Indentation

File: common/mercenary_companies/00_mercenaries.txt [hunk 8]
```diff
@@ -2505,6 +2532,148 @@ wandering_knights_of_rhodes = {
...
+    trigger = {
+		th_is_allowed_to_recruit_mercenaries = yes
+		has_country_flag = keshig_mercenary_company_flag
+	}
+	modifier = { 
+		cavalry_fire = 0.5
+		fire_damage_received = 0.15
+	}
+}
+
```
Comment (Hecatia-Lapislazuli) [inline common/mercenary_companies/00_mercenaries.txt]: Indentation issues - some lines using tab and others spaces

File: common/new_diplomatic_actions/th_challenge_danmaku_duel.txt [hunk 9]
```diff
@@ -1,6 +1,243 @@
...
+			}
+		}
+	}
+	on_accept = {
+		th_create_danmaku_duel_against_country = {
+			country_1 = ROOT
+			country_2 = FROM
+		}
+	}
+    ai_acceptance = {
```
Comment (Hecatia-Lapislazuli) [inline common/new_diplomatic_actions/th_challenge_danmaku_duel.txt]: Indentation

File: common/new_diplomatic_actions/th_challenge_danmaku_duel.txt [hunk 10]
```diff
@@ -1,6 +1,243 @@
...
+			export_to_variable = {
+				variable_name = ai_value
+				value = 1
+			}
+			multiply_variable = {
+				which = ai_value
+				value = -50
+			}
+		}
+    }
```
Comment (Hecatia-Lapislazuli) [inline common/new_diplomatic_actions/th_challenge_danmaku_duel.txt]: Indentation

File: common/scripted_effects/01_scripted_effects_for_on_actions.txt [hunk 11]
```diff
@@ -1788,6 +1836,61 @@ on_trade_good_changed_government_effect = {
...
+			if = {
+				limit = {
+					check_variable = {
+						which = NEDOrangistCauseVar
+						value = 25
+					}
+				}
+				add_ruler_personality = orangist_stadholder_personality
+			}
+        }
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_effects/01_scripted_effects_for_on_actions.txt]: Indentation

File: common/scripted_effects/01_scripted_effects_for_on_actions.txt [hunk 12]
```diff
@@ -1867,6 +1973,46 @@ on_battle_won_unit_effect = {
...
+	}
+	if = {
+        limit = {
+            unit_owner = {
+                has_government_attribute = orangist_cause
+				is_orangists_in_power = yes
+            }
+            is_ruler_commanding_unit = yes
+        }
+	}
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_effects/01_scripted_effects_for_on_actions.txt]: Indentation

File: common/scripted_effects/01_scripted_effects_for_on_actions.txt [hunk 13]
```diff
@@ -2504,6 +2784,19 @@ judaism_on_culture_country_effect = {
...
+        }
+        every_owned_province = {
+            limit = {
+                culture_group = turko_semitic
+                has_owner_accepted_culture = yes
+                has_province_modifier = hsn_ayyubid_tolerance
+            }
+            remove_province_modifier = hsn_ayyubid_tolerance
+        }
+    }
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_effects/01_scripted_effects_for_on_actions.txt]: Indentation

File: common/scripted_effects/th_scripted_effects_for_on_actions.txt [hunk 14]
```diff
@@ -1461,4 +1461,54 @@ th_on_mercenary_recruited_effect = {
 		}
 		THIS = { remove_loot = { who = FROM amount = 250 } }
 	}
+}
+
+th_trigger_on_culture_changed_on_action = {
+	set_province_flag = th_forced_on_action
+	export_to_variable  = {
+		which = th_original_culture
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_effects/th_scripted_effects_for_on_actions.txt:1469]: Isn't actually converted back, which seems is intent.

File: common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt [hunk 15]
```diff
@@ -0,0 +1,778 @@
+th_has_danmaku_duels_unlocked = {
+	custom_trigger_tooltip = {
+		tooltip = th_has_danmaku_duels_unlocked_tt
+		OR = {
+			technology_group = tech_gensokyan_lunarian
+			technology_group = tech_lunarian
+			technology_group = tech_gensokyan
+		}
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt:8]: Use `th_has_touhou_technology = yes`

File: common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt [hunk 16]
```diff
@@ -0,0 +1,778 @@
...
+	}
+}
+
+th_has_not_danmaku_duels_unlocked = {
+	custom_trigger_tooltip = {
+		tooltip = th_has_not_danmaku_duels_unlocked_tt
+		NOT = { technology_group = tech_gensokyan_lunarian }
+		NOT = { technology_group = tech_lunarian }
+		NOT = { technology_group = tech_gensokyan }
```
Comment (Hecatia-Lapislazuli) [inline common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt:17]: Use `th_has_touhou_technology = no`

File: common/scripted_triggers/th_scripted_triggers_for_respawn_feature.txt
Comment (Hecatia-Lapislazuli) [inline common/scripted_triggers/th_scripted_triggers_for_respawn_feature.txt:1]: Dynasties marked as TODO

File: events/TouhouDanmaku.txt
Comment (Hecatia-Lapislazuli) [inline events/TouhouDanmaku.txt:1]: General issues with missing tabs for indentation

File: localisation/touhou_danmaku_l_english.yml
Comment (Hecatia-Lapislazuli) [inline localisation/touhou_danmaku_l_english.yml:1]: Would be good for @r-FlanboysUnite to review, most likely. Seems okay at first glance, though.

File: missions/EMP_German_Missions.txt [hunk 17]
```diff
@@ -295,9 +303,9 @@ emp_germany_hlr_2_ger = {
 			is_in_deficit = no
 			custom_trigger_tooltip = {
 				tooltip = ger_chancellor_tt
-				th_has_mil_advisor_3 = yes
-				th_has_adm_advisor_3 = yes
-				th_has_dip_advisor_3 = yes
+				has_mil_advisor_3 = yes
+				has_adm_advisor_3 = yes
+				has_dip_advisor_3 = yes
```
Comment (Hecatia-Lapislazuli) [inline missions/EMP_German_Missions.txt]: Don't see why they shouldn't be `th_has_x_advisor_3`

**Review by Hecatia-Lapislazuli — CHANGES_REQUESTED** [review]
A lot of inconsistent indentation with switching between tabs and indents, should be addressed by turning them all into tabs, most likely.

File: common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt (hunk 16 above)
Comment (ComradeFlan) [reply common/scripted_triggers/th_scripted_triggers_for_danmaku_fight.txt:17]: Scripted triggers work not great with the no argument

File: common/scripted_triggers/th_scripted_triggers_for_respawn_feature.txt
Comment (ComradeFlan) [reply common/scripted_triggers/th_scripted_triggers_for_respawn_feature.txt:1]: That is intentional, those characters do not have a nation yet associated with a dynasty for them

File: localisation/touhou_danmaku_l_english.yml
Comment (ComradeFlan) [reply localisation/touhou_danmaku_l_english.yml:1]: Fixed

File: common/custom_gui/th_danmaku_gui.txt (hunk 3 above)
Comment (ComradeFlan) [reply common/custom_gui/th_danmaku_gui.txt:183]: Fixed

**Review by Hecatia-Lapislazuli — APPROVED** [review]

**Pull request merged**